main.py

```diff
@@ -829,79 +830,159 @@ def handle_link_down(self, event):
         """Change circuit when link is down or under_mantenance."""
         link = event.content["link"]
         log.info("Event handle_link_down %s", link)
-        switch_flows = {}
-        evcs_with_failover = []
-        evcs_normal = []
-        check_failover = []
-        failover_event_contents = {}
 
-        for evc in self.get_evcs_by_svc_level():
-            with evc.lock:
-                if evc.is_affected_by_link(link):
-                    evc.affected_by_link_at = event.timestamp
-                    # if there is no failover path, handles link down the
-                    # tradditional way
-                    if (
+        with ExitStack() as exit_stack:
+            exit_stack.enter_context(self._lock)
```

Comment:
`handle_link_down` will always be executed single threaded, `self._lock` was meant only for consistency check related execute loop, I wouldn't recommend mixing it here, so until we get to consistency part I wouldn't even touch this. 

Each `evc.lock` though yes, those ones we need to make sure that when acquired by this single threaded exec no other concurrent threads would be performing other side effects. The `ExitStack()` usage is great, if it'll be used, needs to also ensure that deletions are also executed atomically. So, if you can review this.

Reply:
In summary: all affected EVCs need to build/map the flows and perform the flows removals and installations with the bulk request that we discussed in the issue.

Reply:
Alright, I'll change it to a separate lock. A lock is necessary here if we where to have multiple threads that wanted to hold multiple EVC locks, so I'll add one that is named explicitly for such purposes.